File: solutions/clojure/115-basic-calculator-ii.clj
;; Problem 115: Basic Calculator II (LeetCode 227)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn calculate2 [s]
  (let [chars (vec (remove #(Character/isWhitespace %) s))]
    (letfn [(read-num [i acc]
              (if (and (< i (count chars))
                       (Character/isDigit (chars i)))
                (read-num (inc i)
                  (+ (* acc 10) (Character/digit (chars i) 10)))
                [acc i]))
            (parse [i stack op]
              (let [[num next] (read-num i 0)
                    ns (case op
                         \* (conj (pop stack) (* (peek stack) num))
                         \/ (conj (pop stack) (quot (peek stack) num))
                         \- (conj stack (- num))
                         (conj stack num))]
                (if (>= next (count chars))
                  (reduce + 0 ns)
                  (parse (inc next) ns (chars next)))))]
      (parse 0 [] \+))))

File: solutions/typescript/115-basic-calculator-ii.ts
/**
 * Problem 115: Basic Calculator II (LeetCode 227)
 * Difficulty: Med
 * Language: TypeScript
 */
function calculate2(s: string): number {
  const tokens = [...s.replace(/\s/g, '')];
  const readNum = (i: number, acc: number): [number, number] =>
    i < tokens.length && tokens[i] >= '0' && tokens[i] <= '9'
      ? readNum(i + 1, acc * 10 + Number(tokens[i])) : [acc, i];
  const parse = (i: number, stack: number[], op: string): number => {
    const [num, next] = readNum(i, 0);
    const newStack = op === '*'
      ? [...stack.slice(0, -1), stack[stack.length - 1] * num]
      : op === '/'
      ? [...stack.slice(0, -1), Math.trunc(stack[stack.length - 1] / num)]
      : op === '-' ? [...stack, -num] : [...stack, num];
    if (next >= tokens.length) return newStack.reduce((a, b) => a + b, 0);
    return parse(next + 1, newStack, tokens[next]);
  };
  return parse(0, [], '+');
}
